File: backend/routes/categories.py
"""Categories API. Create, Edit, Fetch, and Delete"""

from flask import Blueprint, jsonify, request
import sys
import os

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from imports import *

categories_bp = Blueprint('categories', __name__, url_prefix='/categories')


@categories_bp.route('', methods=['GET'])
def get_categories():
    categories = run_query(FETCH_ALL_CATEGORIES)
    return jsonify([serialize_category(row) for row in categories])


@categories_bp.route('', methods=['POST'])
def add_category():
    name = request.json.get('name', 'New Category')
    category_id = run_query(MAKE_CATEGORY, params=(name,))
    return jsonify({"id": category_id, "name": name}), 201


@categories_bp.route('/<int:category_id>', methods=['PUT'])
def rename_category(category_id):
    data = request.json
    name = data.get('name')
    if not name:
        return jsonify({"error": "Category name is required"}), 400

    run_query(UPDATE_CAT_IN_DB, params=(name, category_id))
    return jsonify({"id": category_id, "name": name}), 200


@categories_bp.route('/<int:category_id>', methods=['DELETE'])
def delete_category(category_id):
    if category_id == 1:
        return jsonify({"error": "Cannot delete default category"}), 400

    run_query(UPDATE_HAB_CATS, params=(category_id,))
    run_query(DEL_CAT, params=(category_id,))
    return jsonify({"message": "Category deleted"}), 200

File: mobile/backend/routes/categories.js
const express = require('express');
const router = express.Router();

const { runQuery } = require('../services/db');
const queries = require('../queries');

router.get('/', async (req, res) => {
  try {
    const rows = await runQuery(queries.FETCH_ALL_CATEGORIES);
    res.json(rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Failed to fetch categories' });
  }
});

router.post('/', async (req, res) => {
  const { name } = req.body;
  try {
    const result = await runQuery(queries.MAKE_CATEGORY, [name || 'New Category']);
    res.status(201).json({ id: result.id, name: name || 'New Category' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Failed to create category' });
  }
});

router.put('/:category_id', async (req, res) => {
  const { category_id } = req.params;
  const { name } = req.body;
  if (!name) {
    return res.status(400).json({ error: 'Category name is required' });
  }
  try {
    await runQuery(queries.UPDATE_CAT_IN_DB, [name, category_id]);
    res.json({ id: Number(category_id), name });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Failed to rename category' });
  }
});

router.delete('/:category_id', async (req, res) => {
  const { category_id } = req.params;
  const defaultCategoryId = 1;
  if (Number(category_id) === defaultCategoryId) {
    return res.status(400).json({ error: 'Cannot delete default category' });
  }
  try {
    await runQuery(queries.UPDATE_HAB_CATS, [defaultCategoryId, category_id]);
    await runQuery(queries.DEL_CAT, [category_id]);
    res.json({ message: 'Category deleted' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Failed to delete category' });
  }
});

module.exports = router;
